File: leet code/Missing Number.cpp


/* 
    Solution by Rahul Surana
    
    ***********************************************************


Given an array nums containing n distinct numbers in the range [0, n], 
return the only number in the range that is missing from the array.



    ***********************************************************
*/



#include <bits/stdc++.h>

class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int x = 0, n = nums.size();
        for(int i = 0; i < nums.size(); i++){
            x = x^nums[i];
        }
        if(n%4 == 3){
            return x;
        }
        else if(n%4 == 2){
            return x^(n+1);
        }
        else if(n%4 == 1){
            return x^1;
        }
        else{
            return x^n;
        }
    }
};

File: leet code/Missing Number.java


/* 
    Solution by Rahul Surana
    
    ***********************************************************


Given a directed acyclic graph (DAG) of n nodes labeled from 0 to n - 1, 
find all possible paths from node 0 to node n - 1 and return them in any order.

The graph is given as follows: graph[i] is a list of all nodes you can visit from 
node i (i.e., there is a directed edge from node i to node graph[i][j]).


    ***********************************************************
*/


class Solution {
    public int missingNumber(int[] nums) {
        int ans = 0, n = nums.length;
        for(int x: nums) ans ^= x;
        if(n%4 == 0) return (ans^n);
        if(n%4 == 1) return (ans^1);
        if(n%4 == 2) return (ans^(n+1));
        return ans;
    }
}